language-plutus-core/src/Language/PlutusCore/TypeRenamer.hs

```diff
@@ -1,120 +1,182 @@
+-- while keeping the table for renaming local (so that we don't rename things in
+-- function applications)
+renameTerm :: Identifiers -> Term TyName Name a -> MaxM (Term TyName Name a)
+renameTerm st t@(LamAbs x (Name x' s (Unique u)) ty t') = do
+    m <- get
+    let st' = modifyIdentifiers u m st
+        pastDef = lookupId u st
     case pastDef of
-        Just _ -> LamAbs x (Name x' s (Unique $ m+1)) <$> rewriteWith (Unique u) (Unique $ m+1) t'
+        Just _ ->
+            modify (+1) >>
+            LamAbs x (Name x' s (Unique (m+1))) <$> renameType st' ty <*> renameTerm st' t'
         _      -> pure t
-renameTerm (TyInst x t tys) = TyInst x <$> renameTerm t <*> traverse renameType tys
-renameTerm (Apply x t ts)   = Apply x <$> renameTerm t <*> traverse renameTerm ts
-renameTerm (Unwrap x t)     = Unwrap x <$> renameTerm t
-renameTerm x                = pure x
-
-cataM :: (Recursive t, Traversable (Base t), Monad m) => (Base t a -> m a) -> (t -> m a)
-cataM phi = c where c = phi <=< (traverse c . project)
-
--- rename a particular unique in a subterm
-rewriteWith :: Unique -> Unique -> Term a -> IdentifierM (Term a)
-rewriteWith i j@(Unique u) = cataM aM where
-    aM (VarF x (Name x' s i')) | i == i' = do
-        modify (first (IM.insert u s))
-        pure $ Var x (Name x' s j)
-    aM (LamAbsF x (Name x' s i') t) | i == i' = do
-        modify (first (IM.insert u s))
-        pure $ LamAbs x (Name x' s j) t
-    aM x = pure (embed x)
-
--- TODO do the same thing here
-renameType :: Type a -> IdentifierM (Type a)
-renameType = pure
+renameTerm st t@(Fix x (Name x' s (Unique u)) ty t') = do
+    m <- get
+    let st' = modifyIdentifiers u m st
+        pastDef = lookupId u st
+    case pastDef of
+        Just _ ->
+            modify (+1) >>
+            Fix x (Name x' s (Unique (m+1))) <$> renameType st' ty <*> renameTerm st' t'
+        _ -> pure t
+renameTerm st t@(Var x (Name x' s (Unique u))) =
+    case pastDef of
+        Just j -> pure $ Var x (Name x' s (Unique j))
+        _      -> pure t
+    where pastDef = lookupId u st
+renameTerm st (Apply x t ts) = Apply x <$> renameTerm st t <*> traverse (renameTerm st) ts
+renameTerm st (Unwrap x t) = Unwrap x <$> renameTerm st t
+renameTerm _ x@Constant{} = pure x
+renameTerm _ x = pure x
+
+renameType :: Identifiers -> Type TyName a -> MaxM (Type TyName a)
+renameType st ty@(TyLam x (TyName (Name x' s (Unique u))) k ty') = do
+    m <- get
+    let st' = modifyIdentifiers u m st
+        pastDef = lookupId u st
+    case pastDef of
+        Just _ ->
+            modify (+1) >>
+            TyLam x (TyName (Name x' s (Unique (m+1)))) k <$> renameType st' ty'
+        _ -> pure ty
+renameType st ty@(TyVar x (TyName (Name x' s (Unique u)))) =
+    case pastDef of
+        Just j -> pure $ TyVar x (TyName (Name x' s (Unique j)))
+        _      -> pure ty
+    where pastDef = lookupId u st
+renameType _ x = pure x
```

Comment:
Again, please no catch-all case.

We also need to handle `TyFix` and `TyForall`.